ironfish/src/wallet/scanner/noteDecryptor.ts

```diff
@@ -0,0 +1,144 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { BufferMap } from 'buffer-map'
+import { Assert } from '../../assert'
+import { BlockHeader } from '../../primitives'
+import { Transaction } from '../../primitives/transaction'
+import { WorkerPool } from '../../workerPool'
+import { DecryptedNote, DecryptNotesOptions } from '../../workerPool/tasks/decryptNotes'
+import { Account } from '../account/account'
+import { Queue } from '../scanner/queue'
+
+// account id -> transaction hash -> decrypted note
+export type AccountTransactionNotesMap = Map<string, BufferMap<Array<DecryptedNote>>>
+
+export type DecryptNotesFromTransactionsCallback = (
+  accountTransactionNotesMap: AccountTransactionNotesMap,
+) => Promise<void>
+
+export class BackgroundNoteDecryptor {
+  private isStarted = false
+
+  private triggerFlushed: (() => void) | null = null
+  private triggerStopped: (() => void) | null = null
+
+  private onFlushed: Promise<void> = Promise.resolve()
+  private onStopped: Promise<void> = Promise.resolve()
+
+  private readonly workerPool: WorkerPool
+  private readonly options: DecryptNotesOptions
+  private readonly decryptQueue: Queue<{
+    promise: Promise<ReadonlyMap<string, ReadonlyArray<DecryptedNote | null>>>
+    transactions: ReadonlyArray<Transaction>
+    callback: DecryptNotesFromTransactionsCallback
+  }>
+
+  constructor(workerPool: WorkerPool, options: DecryptNotesOptions) {
+    this.workerPool = workerPool
+    this.options = options
+    this.decryptQueue = new Queue(Math.max(32 * workerPool.numWorkers, 1))
```

Comment:
This will be roughly 524 MB for a 32 core machine worst case with full blocks.

Reply:
> This will be roughly 524 MB for a 32 core machine worst case with full blocks.

The maximum number of workers is capped at 16 by default, so this would be (32 + 1) * 16 * 0.5 MiB = 264 MiB worst case.